packages/d3fc-axis/src/axisOrdinal.js

```diff
@@ -37,10 +40,117 @@ const axisOrdinal = (orient, scale) => {
         return { offset: [0, base.tickPadding()] };
     };
 
+    const decorateRotation = s => {
+        if (labelRotate) {
+            const { rotate, labelHeight, anchor } = rotation;
+
+            const text = s.select('text');
+            const existingTransform = text.attr('transform');
+
+            const offset = sign() * Math.floor(labelHeight / 2);
+            const offsetTransform = isVertical() ? `translate(${offset}, 0)` : `translate(0, ${offset})`;
+
+            text.style('text-anchor', anchor)
+                .attr('transform', `${existingTransform} ${offsetTransform} rotate(${rotate} 0 0)`);
+        }
+    };
+
+    const measureLabels = s => {
+        const labels = scale.domain();
+
+        // Use a test element to measure the text in the axis SVG container
+        s.attr('font-size', 10).attr('font-family', 'sans-serif');
+        const tester = s.append('text');
+        const labelHeight = tester.text('Ty').node().getBBox().height;
+        const maxWidth = Math.max(...labels.map(l => tester.text(l).node().getBBox().width));
+        tester.remove();
+
+        return {
+            labelHeight,
+            maxWidth,
+            measuredSize: labels.length * maxWidth
+        };
+    };
+
+    const calculateRotation = s => {
+        if (labelRotate) {
+            const { labelHeight, maxWidth, measuredSize } = measureLabels(s);
+
+            // The more the overlap, the more we rotate
+            let rotate;
+            if (labelRotate === 'auto') {
+                const range = scale.range()[1];
+                rotate = range < measuredSize ? 90 * Math.min(1, (measuredSize / range - 0.8) / 2) : 0;
+            } else {
+                rotate = labelRotate;
+            }
+
+            rotation = {
+                rotate: isVertical() ? Math.floor(sign() * (90 - rotate)) : Math.floor(-rotate),
+                labelHeight,
+                maxWidth,
+                anchor: rotate ? labelAnchor() : 'middle'
+            };
+        } else {
+            rotation = {
+                rotate: 0,
+                anchor: isVertical() ? labelAnchor() : 'middle'
+            };
+        }
+    };
+
+    const isVertical = () =>
+        orient === 'left' || orient === 'right';
+    const sign = () => (orient === 'top' || orient === 'left') ? -1 : 1;
+
+    const labelAnchor = () => {
+        switch (orient) {
+        case 'top':
+        case 'right':
+            return 'start';
+        default:
+            return 'end';
+        }
+    };
+
     const base = axisBase(orient, scale, {labelOffset, tickPath});
 
+    base.decorate(s => {
+        decorateRotation(s);
+        decorate(s);
+    });
+
     const axis = (selection) => {
-        return base(selection);
+        if (!rotation) calculateRotation(selection);

```

Comment:
what's the reason behind intercepting the 'render' function to cache the rotation value, rather than just decorating?